File: src/Gas_Station.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {integer[]} gas
    # @param {integer[]} cost
    # @return {integer}
    def canCompleteCircuit(self, gas, cost):
        l = [0 for i in range(len(gas))]
        for i in range(len(gas)):
            l[i] = gas[i] - cost[i]
        tmp, ret = 0, 0
        for i in range(len(gas)):
            tmp += l[i]
            if tmp < 0:
                ret, tmp = (i + 1) % len(gas), 0
        tmp = 0
        for i in range(len(gas)):
            tmp += l[(i + ret) % len(gas)]
            if tmp < 0:
                return -1
        return ret


if __name__ == "__main__":
    solution = Solution()
    print solution.canCompleteCircuit([6, 1, 4, 3, 5], [3, 8, 2, 4, 2])

File: src/algorithms/cpp/Gas_Station.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        vector <int> rec;
        for (int i = 0; i < SZ(gas); ++ i)
            rec.push_back(gas[i] - cost[i]);
        int tmp = 0, ret = 0;
        for (int i = 0; i < SZ(gas); ++ i) {
            tmp += rec[i];
            if (tmp < 0) {
                ret = (i + 1) % SZ(gas);
                tmp = 0;
            }
        }
        tmp = 0;
        for (int i = 0; i < SZ(gas); ++ i) {
            tmp += rec[(i + ret) % SZ(gas)];
            if (tmp < 0)    return -1;
        }
        return ret;
    }
};

int main() {
    Solution solution = Solution();
    vector <int> input1 = {2, 3, 1}, input2 = {3, 1, 2};
    cout << solution.canCompleteCircuit(input1, input2) << endl;
    return 0;
}
